Feature/qa6772 refactor b2b

speratimatteo opened the pull request:
Ripristinati gli step dei destinatari, non più con stringhe ma con costanti dell'Enum dei destinatari (codice più robusto e ripetizioni di codice evitate).
SharedSteps quasi concluso (seguiranno ulteriori micro modifiche)
Inizio rework di AvanzamentoNotifiche B2B steps, seguendo gli stessi principi logici.

speratimatteo committed:
QA6772

Inizio refattorizzazione shared steps (fasi preliminari per capire quali metodi si possono riciclare il più possibile)
Creazione di notifica con una data versione e aggiunta di un destinatario specifico sono sicuramente tra questi

speratimatteo committed:
PN-5947

Creazione Enum per i destinatari

speratimatteo committed:
Merge branch 'feature/5947-fullRefactorWebhookPostV27' into feature/QA6772-fullRefactoSharedSteps

speratimatteo committed:
PN-5947

Continuo refactor, aggiunta di NotificationSteps anche per le versioni 1, 2 e 21

speratimatteo committed:
Merge branch 'develop' into feature/QA6772-fullRefactoSharedSteps

speratimatteo committed:
PN-5947

Mega commit per prima run di prova

speratimatteo committed:
QA 6772

PN-5947

Continuo refactor, seconda run di prova:
- corrette le NullPointer su FullSentNotification,
- Migliorato le costanti
- Rifattorizzato metodo di controllo presenza o meno di searchQ
- Rimosso campi inutili
- altro

speratimatteo committed:
QA6772

Ulteriori piccole modifiche, rimozione codice commentato reso desueto dal nuovo, ottimizzazione di alcuni metodi
Seconda run di prova in dev (quella precedente non era stato pushato il tag)

speratimatteo committed:
QA6772

Fix di alcune null poointer e file feature che erano stati modificati male.
Ultima run in DEV con "soli" 84 errori (e molti per token scaduti). A fronte di queste modifiche mi appresto a lanciare in TEST con un output previsto di massimo 50 errori

speratimatteo committed:
QA 6772

Update Cucumber Analogic nell'Enum Destinatario (il fatto che fosse segnato come PG avendo un codice fiscale alfa numerico causava il fail di alcuni test)

speratimatteo committed:
QA 6772

Piccole migliorie qua e là

speratimatteo committed:
Merge remote-tracking branch 'origin/feature/QA-6733-attivazione-raster-scarto-CON996' into feature/QA6772-fullRefactoSharedSteps

speratimatteo committed:
QA 6772

fix refuso

speratimatteo committed:
QA6772

Ultime modifiche a file feature corrotti in seguito a replace.
Mergiato il branch di Giulia;
Ultima run in test conclusa con 35 errori indipendenti dal codice (problemi radd), me ne aspetto un numero simile

speratimatteo committed:
QA6772

Fix al metodo senza destinatario (che in realtà ne metteva uno, semplicemente senza taxId)... ora invece passiamo effettivamente una request con lista di destinatari vuota.
Fix ad alcuni metodi degli stream che usavano come destinatario Giovanna D'arco (non corretto)
Aggiunta di "nessuno" all'enum dei destinatari

speratimatteo committed:
QA 6772

Fix di alcune regressioni dovute a suggerimenti errati di Sonar
(un campo annotato con @NonNull causava il suggerimento di rimuovere l'if che verificava non fosse null, che però era in realtà necessario)

speratimatteo committed:
QA6772

- Unificati alcuni metodi con struttura simile per l'invio di notifiche che andranno in refused in un unico metodo;
-ripristinati tutti i metodi in B2BUtils, per quanto molti non siano utilizzati, e ordinati per versione.

speratimatteo committed:
QA 6772

Altre migliorie sparse, qualche bug fix

speratimatteo committed:
QA-6772

Correttive anche a AvanzamentoNotificheB2BSteps
Creazione di un interfaccia comune per gestire i pagamenti
Ancora molta pulizia da fare, run preliminare di prova
Eliminati tutti i fullSentNotification con N versioni da SharedSteps (verranno sostituiti da chiamate al b2b tramite IUN)

speratimatteo committed:
QA 6772

Interfaccia (rimasta fuori per errore dal comit precedente)

speratimatteo committed:
QA6772

Fix ad alcune sviste

speratimatteo committed:
Merge branch 'develop' into feature/QA6772-fullRefactoSharedSteps

speratimatteo committed:
QA6772

fix nome variabile

speratimatteo committed:
QA 6772

Ripristino di tutti i file feature dove era stato messo il parametro destinatario come stringa, ora accetta un Destinatario (aggiunto metodo in Destinatario con @ParameterType in grado di restituire l'oggetto Destinatario a partire dal nome).
Anche il metodo addRecipientToNotification di NotificationStepInterface è stato modificato in modo che prenda un destinatario e non più una stringa.
Altre fix minori

speratimatteo committed:
Merge branch 'feature/QA6772-fullRefactoSharedSteps' into feature/QA6772-refactorTotaleAncheB2bSteps

speratimatteo committed:
QA 6772

microfix dovuta a refactor

speratimatteo committed:
QA 6772

Riutilizzo di destinatario anche su step della RADD
Staticizzazione del metodo @ParameterType destinatario (non statico causava il fail dello step)

speratimatteo committed:
QA6772

Piccole correzioni a refusi
aggiunta di alcuni messaggi per InvalidArgumentException

speratimatteo committed:
QA6772

Fix di step in modo che usino il Destinatario (enum)

speratimatteo committed:
QA 6772

Alcune fix a null pointer. Aggiunta B2bStepV21.
Aggiunto case REFINEMENT al wait until timelineElement

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/SharedSteps.java:
```diff
@@ -221,118 +152,73 @@ public class SharedSteps {
-    }
-
-    private HashMap<String, String> mapAllegatiNotificaSha256 = new HashMap<>();
+    /** Anzichè avere 10.000 istanze di FullSentNotification da impostare ogni volta, ora setteremo solo lo IUN della notifica creata.
+     * Dove prima si richiama sharedSteps.getFullSentNotificationX, adesso si recupererà direttamente la versione chiamando il metodo
+     * del B2bClient relativo alla versione che ci interessa, passandogli questo IUN (che è universale) */
```
è possibile migliorare il commento esplicitando soltanto il comportamento attuale, evitando i dettagli sul pregresso che verrà rimosso?

speratimatteo replied:
Risolto

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/SharedSteps.java:
```diff
@@ -221,118 +152,73 @@ public class SharedSteps {
+     * del B2bClient relativo alla versione che ci interessa, passandogli questo IUN (che è universale) */
+    private String notificationIun;
+
+    public enum NotificationVersion {
+        V1(1), V2(2), V21(21), V23(23), V24(24);
+
+        /**
+         * Scopo di questo campo è quello di poter comparare le versioni con < o >
+         * In questo modo si possono aggiungere controlli nel codice per verificare
+         * se un dato Notification Version è antecedente o successivo a un'altra versione
+         */
+        @Getter
+        private final int value;
+
+        NotificationVersion(int value) {
+            this.value = value;
+        }
+    }
```
è possibile portare al di fuori di SharedSteps questa enum?

speratimatteo replied:
Risolto

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/SharedSteps.java:
```diff
@@ -221,118 +152,73 @@ public class SharedSteps {
-    @Setter
-    private FullSentNotificationV24 notificationResponseCompleteV24;
-    private NewNotificationRequestV24 notificationRequestV24;
+    private final Map<NotificationVersion, NotificationStepsInterface> mapOfVersionSteps = Map.ofEntries(
+            Map.entry(V1, new NotificationStepsV1(this)),
+            Map.entry(V2, new NotificationStepsV2(this)),
+            Map.entry(V21, new NotificationStepsV21(this)),
+            Map.entry(V23, new NotificationStepsV23(this)),
+            Map.entry(V24, new NotificationStepsV24(this))
+    );
```
è possibile portare al di fuori questa logica? magari creando un "Provider" che contenga tutta la logica di versionamento e che restituisca il valore atteso

speratimatteo replied:
Risolto

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/SharedSteps.java:
```diff
@@ -910,214 +502,230 @@ public void vieneGenerataUnaNuovaNotificaConUgualePaProtocolNumber() {
+        NotificationStepsInterface notificationStepsInterface = getNotificationStepInterface(notificationVersion);
+        if (paName != null) {
+            setPaAndSenderTaxId(paName);
+        }
+        //TODO MATTEO: un tempo lo stato era sempre ACCEPTED, ora che è parametrico forse la pollingStrategy andrebbe desunta con qualche metodo che si basa sullo stato
```
questo TODO è stato lavorato?

speratimatteo replied:
In realtà allo stato attuale il codice si comporta esattamente come faceva prima: i possibili status erano ACCEPTED o REFUSED (ve n'è un terzo che ho aggiunto io reworkando un po' lo step, ma andava sempre a chiamare il wait for ACCEPTED) ed entrambe le casistiche adottano sempre la polling strategy VALIDATION_STATUS.
E' più un appunto per ricordarci che qualora vi fosse un nuovo status, che adopera una differente polling strategy, sarebbe il caso di fare un metodo ad hoc che desume la strategy a partire dallo status

speratimatteo committed:
Merge branch 'develop' into feature/QA6772-refactorTotaleAncheB2bSteps

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/SharedSteps.java:
```diff
@@ -1181,601 +788,150 @@ public void viewedAllAddress(String user) {
     }
 
-    private void sendNotification(int wait) {
+    // TODO MATTEO TEST: 8 vecchi metodi sono stati mergiati in questo. Il prossimo step sarebbe capire meglio cosa fanno quei metodi di
+    //  utility richiamati nello switch e rimuoverli da B2bUtils (dove c'azzeccano poco, non sono vere utils se vengono richiamate solo qua).
+    //  Alcuni di questi non vengono nemmeno mai richiamati da nessun file feature
+    //  Altra possibile miglioria: sostituire le stringhe delle tipologie d'errore con costanti all'interno della classe Costanti
```
questo TODO è stato processato? Inoltre, è possibile formalizzare il commento?

speratimatteo committed:
Creazione di Enum a parte per StreamVersion e NotificationVersion

speratimatteo committed:
Rework a seguito dei commenti della PR review

speratimatteo committed:
Edit a javadoc

speratimatteo committed:
6772 - rimozione di metodi inutilizzati dall'interfaccia NotificationStepsInterface e le classi che la implementano

speratimatteo committed:
Merge branch 'develop' into feature/QA6772-refactorTotaleAncheB2bSteps

speratimatteo committed:
6772 - fix a step undefined

speratimatteo committed:
6772 - rimozione di commenti e micro modifiche.
Rimosso versionUsed da AvanzamentoNotificheB2bSteps, in modo che coincida sempre con quella di SharedSteps

speratimatteo committed:
6772 - creazione suite di test per solo Webhook V26

speratimatteo committed:
6772 - fix su asyncValidation

speratimatteo committed:
6772 - fix su asyncValidation, getter di lombok sul notificationIun rimossa, istituito metodo omonimo che quando trova lo IUN null effetta la decode in base64 e imposta il risultato come notificationIun (lancia un'eccezione se anche il requestId della response è null

speratimatteo committed:
6772 - rinomina costante da
NOTIFICATION_STATUS_NOT_REFUSED a NOTIFICATION_STATUS_CANCELLED (molto più parlante) e aggiunta versione al decorateErrorMsg

speratimatteo committed:
6772 - aggiunta versionUsed di default (MOST_RECENT) nel costruttore

speratimatteo committed:
Merge branch 'develop' into feature/QA6772-refactorTotaleAncheB2bSteps
